aea/test_tools/acn_image.py

```diff
@@ -0,0 +1,188 @@
+    AEA_P2P_URI_PUBLIC=f"{META_ADDRESS}:9001",
+    AEA_P2P_URI=f"{META_ADDRESS}:9001",
+    AEA_P2P_DELEGATE_URI=f"{META_ADDRESS}:11001",
+    AEA_P2P_URI_MONITORING=f"{META_ADDRESS}:8081",
+    AEA_P2P_ENTRY_URIS=GENESIS_MADDR,
+    ACN_LOG_FILE="/acn/libp2p_node.log",
+)
+
+NODE2: Dict[str, str] = dict(
+    AEA_P2P_ID="cc096b7be575c11d3d3d2f8a9c9be9bd59b351317b2a114ef7e014cc5a92508e",
+    AEA_P2P_URI_PUBLIC=f"{META_ADDRESS}:9002",
+    AEA_P2P_URI=f"{META_ADDRESS}:9002",
+    AEA_P2P_DELEGATE_URI=f"{META_ADDRESS}:11002",
+    AEA_P2P_URI_MONITORING=f"{META_ADDRESS}:8082",
+    AEA_P2P_ENTRY_URIS=GENESIS_MADDR,
+    ACN_LOG_FILE="/acn/libp2p_node.log",
+)
+
+
+class ACNNodeDockerImage(DockerImage):
+    """Wrapper to ACNNode Docker image."""
+
+    uris: List = [
+        "AEA_P2P_URI_PUBLIC",
+        "AEA_P2P_URI",
+        "AEA_P2P_DELEGATE_URI",
+        "AEA_P2P_URI_MONITORING",
+    ]
+
+    nodes = ["bootstrap"]
+
+    def __init__(
+        self,
+        client: DockerClient,
+        config: Dict,
+    ):
+        """
+        Initialize the ACNNode Docker image.
+
+        :param client: the Docker client.
+        :param config: optional configuration to command line.
+        """
+        super().__init__(client)
+        self._config = config
+        self._extra_hosts = {name: "host-gateway" for name in self.nodes}
+
+    @property
+    def tag(self) -> str:
+        """Get the image tag."""
+        return "valory/acn-node:v0.1.0"
+
+    @property
+    def ports(self) -> List[str]:
+        """Ports"""
+        return [self._config[uri].split(":")[1] for uri in self.uris]
+
+    def _make_ports(self) -> Dict:
+        """Make ports dictionary for Docker."""
+
+        return {f"{p}/tcp": (META_ADDRESS, p) for p in self.ports}
+
+    def create(self) -> Container:
+        """Create the container."""
+        container = self._client.containers.run(
+            image=self.tag,
+            command=["--config-from-env"],
+            detach=True,
+            ports=self._make_ports(),
+            environment=self._config,
+            extra_hosts=self._extra_hosts,
+        )
+        return container
+
+    def wait(self, max_attempts: int = 15, sleep_rate: float = 1.0) -> bool:
+        """Wait until the image is up."""
+
+        i, to_be_connected = 0, {self._config[uri] for uri in self.uris}
+        while i < max_attempts and to_be_connected:
+            i += 1
+            for uri in to_be_connected:
+                try:
+                    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                    host, port = uri.split(":")
+                    result = sock.connect_ex((host, int(port)))
+                    sock.close()
+                    enforce(result == 0, "")
+                    to_be_connected.remove(uri)
+                    logging.info(f"URI ready: {uri}")
+                    break
+                except Exception:  # pylint: disable=broad-except
```

Comment:
should `i` be reset to 0 once a connection has succeed? Or was the intention to have `i` as the max_attempts for the whole fleet?

Reply:
You check for them simultaneously, so no need to reset if I understand correctly.